🎉 Source Shopify: migrate `products`, `product images` and `product variants` to `GraphQL BULK`

bazarnov opened the pull request:
## What
Resolving:
- https://github.com/airbytehq/airbyte-internal-issues/issues/7353
- https://github.com/airbytehq/airbyte-internal-issues/issues/7354
- https://github.com/airbytehq/airbyte-internal-issues/issues/7355

## How
- migrated `Products`, `Product Images` and `Product Variants` to `Shopify GQL BULK`
- removed the parent dependency (Product) for the `product images` and `product variants` streams, they should be independent to have the proper migration process
- extended the `query.py` (BULK Queries module) with 3 more classes for `Product`, `ProductImage` and `ProductVariant` queries
- changed the references for the `stream.py` to make the migrated streams inherit from the `IncrementalShopifyGraphQlBulkStream` 
- added 3 test cases to cover the change
- fixed the tests that were affected by the change

## User Impact
There is a BREAKING CHANGE for the `product variants` stream STATE:

**Before:**
```
"id": 112233
```

**Now:**
```
"updated_at": "2024-01-01T01:01:01+00:00"
```

## Can this PR be safely reverted and rolled back?
- [X] YES 💚,

Because the old STATE for `id` based values is still there for the STATE object, we can safely roll back when something is wrong.

bazarnov commented:
@katmarkham Could you please take a look at the user-facing message about the breaking change? Thank you.

maxi297 commented on airbyte-integrations/connectors/source-shopify/source_shopify/streams/streams.py:
```diff
@@ -167,22 +168,16 @@ class MetafieldProducts(IncrementalShopifyGraphQlBulkStream):
-    nested_entity = "variants"
-    # add `product_id` to each nested subrecord
-    mutation_map = {"product_id": "id"}
+class ProductVariants(IncrementalShopifyGraphQlBulkStream):
```
There are a couple of change with that one. For example, `inventory_policy` goes from `deny` to `DENY`, `inventory_management` goes from `shopify` to `SHOPIFY`, etc... should we fix those or else be explicit in the migration guide?

bazarnov replied:
What difference does it create? We had lower-case string values - now they are upper-case, but the schema for these fields hasn't changed. I don't see it as any kind of `breaking` in any scenario. GraphQL just returns some values like this.

bazarnov replied:
Added to the migration-guide

maxi297 commented on airbyte-integrations/connectors/source-shopify/source_shopify/streams/streams.py:
```diff
@@ -167,22 +168,16 @@ class MetafieldProducts(IncrementalShopifyGraphQlBulkStream):
-    nested_entity = "images"
-    # add `product_id` to each nested subrecord
-    mutation_map = {"product_id": "id"}
+class ProductImages(IncrementalShopifyGraphQlBulkStream):
```
This streams has a lot less data in it

Also, some fields changes:
* 'created_at' goes from '2023-04-14T03:34:46-07:00' to '2023-04-14T10:34:46+00:00' which means we lose the timezone information
* position is missing
* variant_ids goes from [] to not being in the record
etc...

Should we document those and be explicit about it in the migration guide?

bazarnov replied:
> 'created_at' goes from '2023-04-14T03:34:46-07:00' to '2023-04-14T10:34:46+00:00' which means we lose the timezone information

GraphQL omits the `timezone` completely (UTC Only) we cannot do anything about it. The API returns the `2023-04-14T10:34:46Z` date-time string, which we parse to be `rfc3339` in the end, to preserve the `format` for the STATE and not having the breaking changes everywhere.

> position is missing

yes, it's not available for GraphQL.

> variant_ids goes from [] to not being in the record

yes, it's not available for GraphQL.

> Should we document those and be explicit about it in the migration guide?

I'll add the notice about it to the migration guide.

bazarnov replied:
Added to the migration-guide.

maxi297 reviewed (commented):
A couple of concerns:
* There are records missing in `product_images` during CATs execution. Is it expected? 
* Should we run a couple of live-testing tests on that to ensure there are no regressions for users as well? It seems like there are a couple of differences we might want to address or be explicit about it

I haven't check the products stream but I expect live testing would help here as well

bazarnov commented on docs/integrations/sources/shopify-migrations.md:
```diff
@@ -1,5 +1,17 @@
 # Shopify Migration Guide
 
+## Upgrading to 2.1.0
+This version implements `Shopify GraphQL BULK Operations` to speed up the following streams:
```
This is not a breaking change because we `added` a new property: `presentment_prices` that contains the `compare_at_prices` as the `object` with `amount` and `currency_code` nested properties. 

The `presentment_prices` was never in the `REST` output, but the schema had an invalid reference though, I've corrected it up to the docs.

The original `compare_at_prices` didn't change (at the root of the schema); it's still a `string.`

maxi297 replied:
I'm confused: [this line](https://github.com/airbytehq/airbyte/pull/37767/files#diff-93f120238b8f5d9fda0d64fcfb955b38f71fae919523c262e5f4731a332ab3d9R124) seems to indicate to me that it was previously a number and is now an object. What am I missing?

bazarnov replied:
If you look at the old `expected_records,`, you will not find the `presentment_prices` included as additional proof.

maxi297 replied:
I'm not sure I follow. Can you give me an example of value for `compare_at_prices` before the change? Our expected records do not populate this value

bazarnov replied:
I'll add the description about it to the migration guide.

bazarnov commented on airbyte-integrations/connectors/source-shopify/integration_tests/expected_records.jsonl:
```diff
@@ -70,9 +70,8 @@
 {"stream": "price_rules", "data": {"id": 1112171741373, "value_type": "fixed_amount", "value": "-10.0", "customer_selection": "all", "target_type": "line_item", "target_selection": "all", "allocation_method": "across", "allocation_limit": null, "once_per_customer": false, "usage_limit": null, "starts_at": "2017-01-19T09:59:10-08:00", "ends_at": null, "created_at": "2022-10-14T10:19:39-07:00", "up ...
 {"stream": "price_rules", "data": {"id": 945000284349, "value_type": "percentage", "value": "-3.0", "customer_selection": "all", "target_type": "line_item", "target_selection": "all", "allocation_method": "across", "allocation_limit": null, "once_per_customer": true, "usage_limit": 10, "starts_at": "2021-07-07T07:22:04-07:00", "ends_at": null, "created_at": "2021-07-07T07:23:11-07:00", "updated_a ...
 {"stream": "price_rules", "data": {"id": 945205379261, "value_type": "percentage", "value": "-100.0", "customer_selection": "all", "target_type": "shipping_line", "target_selection": "all", "allocation_method": "each", "allocation_limit": null, "once_per_customer": false, "usage_limit": null, "starts_at": "2021-07-08T05:40:13-07:00", "ends_at": "2024-01-01T23:59:59-08:00", "created_at": "2021-07- ...
-{"stream": "product_images", "data": {"id": 29301295481021, "alt": null, "position": 1, "product_id": 6796218138813, "created_at": "2021-06-22T18:09:28-07:00", "updated_at": "2021-06-22T18:09:28-07:00", "admin_graphql_api_id": "gid://shopify/ProductImage/29301295481021", "width": 4393, "height": 2929, "src": "https://cdn.shopify.com/s/files/1/0580/3317/6765/products/tin-of-beard-balm.jpg?v=162441 ...
```
Because we now rely on the different cursor, thus the records are filtered out not by the `id,` but by the `updated_at,` therefore, the old ones are filtered out from the output. If you check the `updated_at` for the old records in the file, you will see non-relevant to the `start_date` records, which should not be emitted in an ideal world.

Long story short: now the `product_images` are fetched using the proper server-side filtering.

maxi297 replied:
I'm not sure I buy into that. The sync I've done was a full refresh. My understanding is that it's also the sync mode that CATs use. My reasoning is that cursor field and states shouldn't influence the records returned during full_refresh sync. Does that make sense? Or have you done changes that makes it that both `latest` and this branch now return the same records?

maxi297 replied:
Ok, that makes sense. So basically, this stream is now subject to the `start_date` which if provided as `"start_date":"2023-04-13"` in the config. Since those records are from 2021, they are not picked up anymore. I'll resolve this comment

bazarnov commented on airbyte-integrations/connectors/source-shopify/source_shopify/streams/base_streams.py:
```diff
@@ -648,7 +648,7 @@ def __init__(self, config: Dict) -> None:
         self.bulk_window_in_days = config.get("bulk_window_in_days")
         if self.bulk_window_in_days:
-            self.job_manager.job_size = self.bulk_window_in_days
+            self.job_manager._job_size = self.bulk_window_in_days
```
This is the small regression, after the latest PR, so I've corrected it to be the `public` because we use it more than once in the CDK methods.

bazarnov commented:
> * There are records missing in `product_images` during CATs execution. Is it expected?

Yes, because previously, it was based on the `Id` cursor, and now `product_images` uses the `updated_at` cursor, which captures the records using the `updated_at` field. (When the parent dependency (Product) was present, we fetched all the products and filtered not by `updated_at`, but by the `id`)

> Should we run a couple of live-testing tests on that to ensure there are no regressions for users as well? It seems like there are a couple of differences we might want to address or be explicit about it

These will be added in a separate PR. I don't have a clear path to implementing live tests for the Shopify source. For now, the unit_tests cover the bulk flow, and no regressions are faced. I've extended the unit_tests to cover a couple of additional cases we want to be retryable and removed the `patch` logic.

bazarnov commented:
Update the migration-guide with the `Products` stream changes as well.

maxi297 reviewed (commented):
Should we run live testing on a couple of connections to validate the changes?

bazarnov commented:
Some intermediate `Live Tests` results:

```markdown
## Tested Connection IDs:
1)  ID: `80be7c99-7a87-41e9-bae9-b0e61e7be4c4`
2)  ID: `89ed6b0d-3cdf-46e1-bbee-35061c038471`
```

Results
ID: `80be7c99-7a87-41e9-bae9-b0e61e7be4c4`
[GH Action](https://github.com/airbytehq/airbyte/actions/runs/9003718898)
[Report](https://ff71b578b8e31ffe609a9dce5e97d3dc4d4163aca97900b24f009fe-apidata.googleusercontent.com/download/storage/v1/b/airbyte-ci-reports-multi/o/airbyte-ci%2Fconnectors%2Ftest%2Fmanual%2Fbaz_source_shopify_migrate-product-images-variants-to-bulk%2F1715179066%2F805cd2979df3e30bd900171ba85d6b3fa6dfd71d%2Fsource-shopify%2F2.1.0%2Foutput.html?jk=ASOpP9jDP4kkleGGYvoymmDBn32hRp0N4KP7eE8k51sOneE8vm5bCLUZLBhhvD8WftVfDPZGcOXfj7Qxr_cyxfKZH44aZy6NfSDzOlR2E-nLWXDDAtbI7hX8QncbjfGabpw7OLjvpWSbGKa5GAS1uE3Qj13utdFijQ6uhw-5sh5FR_Vk1RM4ugJK9yaaX-cOtzWA0td9Sva-ns_bk6h4wFzE-FyHy9XF5MqqNpFTkksrIZxXGoYrhPlMJwp6x-cC2KHQI_ZSA6utesSq4JGJcRdSfqRluy5i9ZNYA1jjB1kqMlSm1iUv4vOqS98TDch5XS9Udond61px3-PRSzZEa6khFCHc3bfTBxQxKZNpEyGOkMzOdg3x1LGq2aAY8vDjxzGWRmzv8LhPpauFhervMwzzGcQEY-vrhvzGVyA_37ec39T9iHx3ZVQRGwUf3tzhCBqf9qfTsY9jbkO_zIgQkxoDsCBpwySanxS1OwAhYxJI-6_jPAGLkqc_gXADdH6Kez0va-L9RFFtka-BUujgKhzjufxJrTP1JgXAFx7cpWvwCsLh9O1-diqkEUBjPRVF4n26rT0BJNruhk1pTBmyYghz9mTXawvWiWE_1-39xdQtr_hfwy1mXuVQw-tT8d_-PMGPPgvLa0uYctUMmexfsjq27LGbAsEhewwJPyVcU_pvHmi8ae9k1rLty-JGtL0xqMCC9maIQSsXauvwqLig4SXdZBSlnH-KRZqfO5vRbjKWrz3hXpLC1SyUJ27YMLrQjEFIKLUzchwJMp9uN9_DbYmgvjNLmH0SjlOqSef_0MOF09SvfUBqt7Bz5vMgqjJp_Q2oS_WFIzKdGGgaaKrokF5VRxf5A_WF3ap_-kjmvZROMRJuUzMScmZ57rc_4jd-77uwD9wSBjJwlDlJgff5En25ZmCGgBa9nmLg3obDcJSeZCMnYcVSrci0lIZyVUbP_2lDGN2wq5wRWIwuzbg73NqbUAB_ddbPX6H-zr2A9EwXrpe9KcUNVL4sPOGW4-YfQNglBndHnq9KRpmZKpSaJ9RBT4kPYXXTVF77Yg4SsAjSnrueW8eJxoBUzArxiZk46rcBuoEsKij3En6ZAf0dcxH7YwvueJjFtz0fTfoWaSOqxWFNXnHAyByMGBFCO62jihJZBitpXKh2Ef0NzOHPFKO-mgmaV9VcTfQS0eZkfY0da135eNu5NuHIy7qqWfvHB1oSNeXk1brh7znPd6zLlRpl6DYmB3gpVhrCn1i0m6xXoQZe38eohvNwjl00IvEIRetWK9LjMNbnxdTnCVgKmZfI9nHYtnxD-NpkMuI5e3TIWazBMG8GfTeO5OMKL9xplCbm-gAFCQ2LsRpJd-oIcU7DyT_LODUXI14VndmRH7IM7zfE0w3CeQ&isca=1)
[Artifact](https://github.com/airbytehq/airbyte/actions/runs/9003718898/artifacts/1484505919)

ID: `89ed6b0d-3cdf-46e1-bbee-35061c038471`
[GH Action](https://github.com/airbytehq/airbyte/actions/runs/9003948004)
[Report](https://ff07a57a4cadc3e165cb64b22ad0293acc844be12f7689e42f8438a-apidata.googleusercontent.com/download/storage/v1/b/airbyte-ci-reports-multi/o/airbyte-ci%2Fconnectors%2Ftest%2Fmanual%2Fbaz_source_shopify_migrate-product-images-variants-to-bulk%2F1715180037%2F805cd2979df3e30bd900171ba85d6b3fa6dfd71d%2Fsource-shopify%2F2.1.0%2Foutput.html?jk=ASOpP9gVsx3T9v4Ikmgiy9elzQg_F6ueq3tDDB7aKVoUVf2R0F1kCd3QYGWZevZbVwNGvrktBW4QmkSKdgTS4NhgcYvp901sQufPm9Ky1a5xMJgmzKKnuQzFMlj3YCi1DwhT-wW5cWu4epsB68ltNMmR_T5HF_d_MnPlTHYIUgOzQLqVymoYDMzHVxo2GEjFsNxkBb9eeK1Bsonbfzahwc6Vjz50AEv9OXJZ9bVie3rQo-MZBLs2vGU5-TAg8rRpmGiI1OHQQm7Wzq_c-nYFV6YLR-6sIIdOuKqv7maEEHI-Zqz5FeaW35F0qCmnl5UaMimSsXeuLDRVzTUDdPHrlIve2zjogn1QwkcxePvnZ-mvxUpxfNx54lyuyzR3sVHARaZ5Ns8l4q6_-x9GQvLwpYTMSUbVhjRuFOXV5YGeXaQFXcXQXz8JkVzQD90dWrFgl1bBQ6fhcwIkfGg8pd5oeapP-rWG5ea5oaCnQ72Ud_4901GLih0XwVPI-MOm8v5uwlBPBwAM1VUV2e2J2aaPcQ132yi-bmCvBizUmsoouGW9qHVdD_z2b0pvknniitU4bubahnt77RGLfXfnelmuV49_zGEpDX7xjnnwARy1dfAzpHmQI1Yz9ewWEszMMVI4c-wg4TEKe2Wgld5kcgs1FRXEIrWqr2vXdnAANc0qmKKju5nhq9MOBUiyqLviiip5sXB_dK3Ca2d-AlhIenwY3__gG-uf-y00r_MRrmQOfHyZ8MOsZr1FuWn_jvI1_nCMh8u71hLxCYxTTUTWBsR8o_j7btfz_lh0X6nPM-y6Wduf6NFTR0qkjEljwTo-ODMJ5Ob35rL5potFT_OqyM1IjPIm2u2BfmNuSH-xlClMl0N8JC3aS5U9_R68zQkXB7aTpn4aXnd_eTl0KbueUD8wezmbyznbqrzV8_EMJUBN_cQ5bSz3iYha9PEZgcPhipZIVoMrMn1CYjlLDnQxmHQnkvnhFLtlUfdHgNjsEgyhn8X4NiCzBIzw8csp_P7ncr34gG0Hbzyx7jjZuCN1wgj0u9r9cKTTRdZ_HDXIRieTsvA11L_9XEbhxvONMRsAi6yWwTMfQVBvguXoQdj-NL2Ow5eQ6zk4nhsNI9uIZnaTWW_MEbSYPLUY8WSzaArqrTS4-d9SGxijqtLFItFF1hSzL6YPkat6X6mwqDDZtdWLoaW6BFLV5gENDpzPrcxWdQE9vrXufe9GY8Tp53tDOOU_CSTasreE2HCCG3RAkS9LqbNEOEuuMfk2cShvmFD7MoHp0CTHOPflnbjejDwIUkTwAJvTwBDMtDOE4coPjs4WYqKfw7djVqNCir_PAurA8Pa3g6CaNSnzt8iDl9_zbNclNWrI0XNhCL3w1EHLnqwXueW2BSnGFFcQTA&isca=1)
[Artifact](https://github.com/airbytehq/airbyte/actions/runs/9003948004/artifacts/1484573623)

bazarnov commented:
The `pre-release` version has been published successfully: `2.1.0-dev.805cd2979d`
Action run: https://github.com/airbytehq/airbyte/actions/runs/9014252053

bazarnov commented:
`Live Tests` results:

## Tested Connection IDs:
```markdown
1) `Products` stream:
- `1adee9d6-d8b1-4425-85a7-e13aebf7331e`
- `80be7c99-7a87-41e9-bae9-b0e61e7be4c4`

2) `Product Images` stream:
- `d82df6a0-1b94-4608-9e63-e9f0d57ec652`
- `2fa9f9ea-5a58-4628-b5f1-08314bfeacaf`

3) `Product Variants` stream:
- `81a5acbb-4374-497d-9af2-297ea8e7f6a7`
- `a7d4dcfe-39bb-4803-8e49-18ae07e838b2`
```

## Results:
```markdown
### `Products` stream:
1)  ID: `1adee9d6-d8b1-4425-85a7-e13aebf7331e`
[GH Action](https://github.com/airbytehq/airbyte/actions/runs/9030841455)
[Artifact]()

2) ID: `80be7c99-7a87-41e9-bae9-b0e61e7be4c4`
[GH Action](https://github.com/airbytehq/airbyte/actions/runs/9030851818)
[Artifact]()

### `Product Images` stream:
1) ID: `d82df6a0-1b94-4608-9e63-e9f0d57ec652`
[GH Action](https://github.com/airbytehq/airbyte/actions/runs/9030861383)
[Artifact]()

2) ID: `2fa9f9ea-5a58-4628-b5f1-08314bfeacaf`
[GH Action](https://github.com/airbytehq/airbyte/actions/runs/9030865668)
[Artifact]()

### `Product Variants` stream:
1) ID: `81a5acbb-4374-497d-9af2-297ea8e7f6a7`
[GH Action](https://github.com/airbytehq/airbyte/actions/runs/9030873645)
[Artifact]()

2) ID: `a7d4dcfe-39bb-4803-8e49-18ae07e838b2`
[GH Action](https://github.com/airbytehq/airbyte/actions/runs/9030879979)
[Artifact]()
```

cc @maxi297

maxi297 commented:
I feel uneasy to accept those live testing results without explaining those gaps:

#### `1adee9d6-d8b1-4425-85a7-e13aebf7331e` and `80be7c99-7a87-41e9-bae9-b0e61e7be4c4`
The HTTP requests are still all the same i.e. I still don't see any request to products.json in the control version which makes me wonder if the stream is tested at all... Can you explain why we don't see HTTP requests to `/products.json` 


#### `d82df6a0-1b94-4608-9e63-e9f0d57ec652`
It says `product_variants stream: records with primary key in target & control whose values differ` but the diff does not highlight [this change](https://github.com/airbytehq/airbyte/pull/37767#discussion_r1588459348). I would also expect to see this discrepancy in `81a5acbb-4374-497d-9af2-297ea8e7f6a7` and `a7d4dcfe-39bb-4803-8e49-18ae07e838b2`. Do we know why this change is not picked up by the live testing?

bazarnov commented:
Because of the inability to test using `live-testing` tool (permissions issues), the manual changes report was conducted.

## CHANGES (REGRESSION) TEST PLAN
1) Make common `CONFIG`
2) Make common `CATALOG` with isolated target (affected) stream
3) Run-Test each affected stream against the `Test Criteria`. Affected streams: `Products`, `Product Images`, `Product Variants`

### COMMANDS:
#### FULL-REFRESH
```bash
## CONTROL

docker run --rm -v $(pwd)/secrets:/secrets -v $(pwd)/integration_tests:/integration_tests airbyte/source-shopify:latest read --config /secrets/config.json --catalog /integration_tests/configured_catalog.json --debug > control.json
```

```bash
## TARGET

docker run --rm -v $(pwd)/secrets:/secrets -v $(pwd)/integration_tests:/integration_tests airbyte/source-shopify:dev read --config /secrets/config.json --catalog /integration_tests/configured_catalog.json --debug > target.json
```

#### INCREMENTAL
```bash
## CONTROL

docker run --rm -v $(pwd)/secrets:/secrets -v $(pwd)/integration_tests:/integration_tests airbyte/source-shopify:latest read --config /secrets/config.json --catalog /integration_tests/configured_catalog.json --state /integration_tests/control_state.json --debug > control_inc.json
```

```bash
## TARGET

docker run --rm -v $(pwd)/secrets:/secrets -v $(pwd)/integration_tests:/integration_tests airbyte/source-shopify:dev read --config /secrets/config.json --catalog /integration_tests/configured_catalog.json --state /integration_tests/target_state.json --debug > target_inc.json
```

### Test Criteria:

1) Outgoing API request example:
    * 1.1) `CONTROL`
    * 1.2) `TARGET`
    * Explain the diff

2) Sync mode: `Full-Refresh`
    * 2.1) `Number of records`:
        * 2.1.1) `CONTROL` 
        * 2.1.2) `TARGET`
        * Explain the diff

    * 2.2) `STATE` comparison:
        * 2.2.1) `CONTROL`
        * 2.2.2) `TARGET`
        * Explain the diff

    * 2.3) `Missing records`:
        * 2.3.1) `CONTROL`
        * 2.3.2) `TARGET`
        * Explain the diff

3) Sync mode: `Incremental`
    * 3.1) `Number of records`:
        * 3.1.1) `CONTROL`
        * 3.1.2) `TARGET`
        * Explain the diff

    * 3.2) `STATE` comparison:
        * 3.2.1) `CONTROL`
        * 3.2.2) `TARGET`
        * Explain the diff

    * 3.3) `Missing records`:
        * 3.3.1) `CONTROL`
        * 3.3.2) `TARGET`
        * Explain the diff

4) `Record example`:
    * 4.1) `CONTROL`
    * 4.2) `TARGET`

5) `Schema changes`:
    * 5.1) Added fields
    * 5.2) Changes types
    * 5.3) Removed fields
    * Explain the diff

6) Is this a `breaking change`:
    * [] Yes
    * [] No


## Test Results ( PII - free ):

- [`Products` stream](https://github.com/airbytehq/airbyte/files/15293674/Products_stream_changes.md)
- [`Product Images` stream](https://github.com/airbytehq/airbyte/files/15293648/Product_Images_stream_changes.md)
- [`Product Variants` stream](https://github.com/airbytehq/airbyte/files/15293677/Product_Variants_stream_changes.md)

cc @maxi297

maxi297 commented on docs/integrations/sources/shopify-migrations.md:
```diff
@@ -1,5 +1,26 @@
+* In the `Product Variants` stream, the `presentment_prices.compare_at_price` property has changed from a `number` to an `object of strings`. This field was not populated in the `REST API` stream version, but it is correctly covered in the GraphQL stream version.
+* The `Product Variants` stream's `inventory_policy` and `inventory_management` properties now contain `uppercase string` values, instead of `lowercase`.
+* In the `Product Images` stream, the date-time fields, such as `created_at` and `updated_at`, now use `UTC` format without a timezone component.
+* In the `Product Images` stream, the `variant_ids` property is no longer available. Refer to the `Product variants` stream instead.
```
I think `position` is also missing based on the output you provided, right?

bazarnov replied:
It's not present in the record for the TARGET version, but schema didn't change. I'll add this mention.

bazarnov replied:
Corrected.

matiasferrin replied:
Hello, position is not missing but it always returns null values. Could you please investigate?

maxi297 commented on airbyte-integrations/connectors/source-shopify/source_shopify/schemas/product_variants.json:
```diff
@@ -121,8 +121,17 @@
             }
```
Based on the sample you provided, the control version has field `image_id` while the target has `image`. Is this expected? Should we document this as well?

bazarnov replied:
No, added the handling of that, we basically need to have `image_id` at the root level, but `graphql` nests it in `"image":{"image_id": "asdfasdf"}`. We don't have this populated for our sandbox, therefore it was `Null` for the data and wasn't processed by far. Thanks for the notice! 

Corrected. Updated the dedicated `unit_test` to reflect this and updated the branch.

maxi297 reviewed (approved):
Shit Baz! Big thumbs up on documenting the changes for this PR. I really appreciate it as it help review and go back to those changes so that we understand why it was done like that. Kudos! 

Based on the report:

> The difference in records is explained by the fact we don't have the possibility to fetch the `deleted` records, using `GraphQL`. The `deleted` records, are available using `REST` only.

Should we add that to the migration notes?

loris commented:
@bazarnov @maxi297 quick question: since there are breaking changes, it should have been included in a 3.0 release no? I can't find how Airbyte Cloud handles connector upgrade, is this automated?

maxi297 commented:
Hi @loris ! The breaking changes were introduced as part of 2.1.0. As this is a breaking change, the opt-in mechanism should kick in and you should see a prompt asking you to upgrade before 2024-06-10

loris commented:
@maxi297 Thanks for the clarification, I'm in discussion with cloud support because this is clearly not working as expected. In our production workspace, connections have been upgraded to 2.2 without any prompt (and thus broke our pipeline), while on our other workspaces, connections are still running Shopify 2.0 connector and we don't see any Upgrade prompt popping up.

loris commented:
@maxi297 Actually I might have found the issue:

```

      2.1.0:
        message:
          "This upgrade changes the `Products`, `Product Images` and `Product Variants` streams to use `Shopify GraphQL BULK`.
          More details here: https://github.com/airbytehq/airbyte/pull/37767."
        upgradeDeadline: "2024-06-10"
        scopedImpact:
          - scopeType: stream
            impactedScopes: ["product_variants"]
```

The breaking change message only states `product_variants` in the `impactedScopes`, while `products` was also clearly impacted. Our Airbyte connections only needed to sync `products` (as the variants where part of that stream too). But with the change, we lost the variant data and were not informed of the breaking change because the `impactedScopes` did not included `products`, does this seem right?

bazarnov commented:
@loris This message included the `breaking change` that breaks the source sync operations only, the info about the data was re-arranged based on the API changes are not included in this message. 

In the migration guide here: https://github.com/airbytehq/airbyte/blob/e9e6cf3778caf804bc5d433ccda1530de1373a6d/docs/integrations/sources/shopify-migrations.md which comes with this release, we included all the details about the actual data change we discovered. Have you had a chance to observe this in the UI?

Hope that answers your question and thank you for your feedback.

loris commented:
Very clear!

> Have you had a chance to observe this in the UI?

This is the part which is confusing to me: No, I did not observe anything in the UI. I also have a workspace where Shopify connector are still running the 2.0 version and did not see any Upgrade banner. I have the email/webhook notifications set up, but nothing there too.

maxi297 commented:
This information will be helpful for us a well as maybe they is a gap we are not aware related to the opt-in mechanism. Please, keep us up to date on this one

loris commented on airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/query.py:
```diff
@@ -1502,3 +1504,483 @@ def record_process_components(self, record: MutableMapping[str, Any]) -> Optiona
+class Product(ShopifyBulkQuery):
+    """
+    {
+        products(query: "updated_at:>='2020-01-20T00:00:00+00:00' AND updated_at:<'2024-04-25T00:00:00+00:00'", sortKey:UPDATED_AT) {
```
@bazarnov Hi, out of curiosity, is this normal to have hard-coded updated_at values here?

bazarnov replied:
For the 'docstring' query example, I believe - yes. 

Does this look wrong to you? 
Should we have dynamic docstring for such cases?

If you can contribute to have the examples like this constructed dynamically, we would appreciate such improvements. 

Thank you for your question, and feel free to add more.
